googet_check.go

```diff
@@ -0,0 +1,119 @@
+
+	"github.com/google/googet/v2/client"
+	"github.com/google/googet/v2/googetdb"
+	"github.com/google/googet/v2/goolib"
+	"github.com/google/googet/v2/install"
+	"github.com/google/googet/v2/system"
+	"github.com/google/logger"
+	"github.com/google/subcommands"
+)
+
+type checkCmd struct {
+	sources string
+	dryRun  bool
+}
+
+func (*checkCmd) Name() string     { return "check" }
+func (*checkCmd) Synopsis() string { return "check and take over exsiting packages" }
+func (*checkCmd) Usage() string {
+	return fmt.Sprintf(`%s check [-sources repo1,repo2...] [-dry_run=true]`, filepath.Base(os.Args[0]))
+}
+
+func (cmd *checkCmd) SetFlags(f *flag.FlagSet) {
+	f.BoolVar(&cmd.dryRun, "dry_run", false, "Don't make any changes to the DB.")
+	f.StringVar(&cmd.sources, "sources", "", "comma separated list of sources, setting this overrides local .repo files")
+}
+
+func (cmd *checkCmd) Execute(ctx context.Context, f *flag.FlagSet, _ ...interface{}) subcommands.ExitStatus {
+	exitCode := subcommands.ExitFailure
+	cache := filepath.Join(rootDir, cacheDir)
+	var filteredState []goolib.RepoSpec
+	db, err := googetdb.NewDB(filepath.Join(rootDir, dbFile))
+	if err != nil {
+		logger.Fatal(err)
+	}
+	defer db.Close()
+	state, err := db.FetchPkgs()
+	var newPkgs client.GooGetState
+	downloader, err := client.NewDownloader(proxyServer)
+	if err != nil {
+		logger.Fatal(err)
+	}
+	repos, err := buildSources(cmd.sources)
+	if err != nil {
+		logger.Fatal(err)
+	}
+	rm := downloader.AvailableVersions(ctx, repos, cache, cacheLife)
+	unmanaged := make(map[string]string)
+	fmt.Println("Searching for unmanaged software...")
+	for r, repo := range rm {
+		fmt.Println(r)
+		for _, p := range repo.Packages {
+			match := false
+			for _, k := range state {
+				if p.PackageSpec.Name == k.PackageSpec.Name {
+					match = true
+					break
+				}
+			}
+			if match {
+				continue
+			}
+			filteredState = append(filteredState, p)
+			app, _ := system.AppAssociation(p.PackageSpec.Authors, "", p.PackageSpec.Name, filepath.Ext(p.PackageSpec.Install.Path))
+			if app != "" {
+				unmanaged[p.PackageSpec.Name] = app
+				if cmd.dryRun {
+					continue
+				}
+				pi := goolib.PackageInfo{
+					Name: p.PackageSpec.Name,
+					Arch: p.PackageSpec.Arch,
+					Ver:  p.PackageSpec.Version,
+				}
+				if err := install.FromRepo(ctx, pi, r, cache, rm, archs, &newPkgs, true, downloader); err != nil {
+					logger.Errorf("Error installing %s.%s.%s: %v", pi.Name, pi.Arch, pi.Ver, err)
+					exitCode = subcommands.ExitFailure
+					continue
+				}
+				logger.Infof("Unmanaged software added to googet database(packagename: application name): %v: %v\n", p.PackageSpec.Name, app)
+			}
+		}
+	}
+	if len(newPkgs) != 0 {
+		if err = db.WriteStateToDB(newPkgs); err != nil {
+			logger.Fatal(err)
+		}
+	}
+	if len(unmanaged) > 0 {
+		fmt.Println("Found the following unmanaged software (Package: Software name) ...")
+		for k, v := range unmanaged {
+			fmt.Printf(" %v: %v\n", k, v)
+		}
+	}
```

Comment:
Output something for the `len(umanaged) == 0` case to bookend the "Searching for unmanaged software..." string.